File: fsharp/grains/Grains.fs
module Grains

let square (n: int): Result<uint64,string> = 
    if n <= 0 || n >= 65 
    then Error "square must be between 1 and 64"
    else Ok (1UL <<< (n-1))

let total: Result<uint64,string> = 
    Ok System.UInt64.MaxValue

File: clojure/grains/src/grains.clj
(ns grains
  (:use [clojure.core]))

(defn square [n]
  (if (<= n 1)
    1
    (*' 2 (square (dec n)))))

;; Can be done more succinctly
;; (defn total
;;   "Finite geometrical series sum."
;;   []
;;   (/ (- 1 (square 65))
;;      (- 1 (square 2))))

;; Based on the observation that 2^0 + 2^1 + ... + 2^n = 2^(n+1) - 1
(defn total []
  (dec (square 65)))
